File: src/main/Java/dfs/boj2644.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.stream.IntStream;

public class boj2644 {private static ArrayList<Integer>[] linked;
    private static boolean[] isVisited;
    private static int personA;
    private static int personB;
    private static int count = 0;
    private static StringBuilder  builder = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        personA = Integer.parseInt(tokenizer.nextToken());
        personB = Integer.parseInt(tokenizer.nextToken());
        linked = new ArrayList[n + 1];
        IntStream.range(1, n + 1).forEach(index -> linked[index] = new ArrayList<>());
        isVisited = new boolean[n + 1];
        int m = Integer.parseInt(reader.readLine());
        while (m-- > 0) {
            tokenizer = new StringTokenizer(reader.readLine());
            int x = Integer.parseInt(tokenizer.nextToken());
            int y = Integer.parseInt(tokenizer.nextToken());
            linked[x].add(y);
            linked[y].add(x);
        }
        dfs(personA);
        if (builder.isEmpty()) writer.append("-1").close();
        else writer.append(builder).close();
    }
    private static void dfs(int people) {
        if (isVisited[people]) return;
        isVisited[people] = true;
        if (people == personB) {
            builder.append(count);
            return;
        }
        count++;
        linked[people].forEach(nextPeople -> dfs(nextPeople));
        count--;
    }
}

File: src/main/Kotlin/dfs/boj2644.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import java.util.stream.IntStream

private var linked = arrayOf(ArrayList<Int>())
private var isVisited = booleanArrayOf()
private var count = 0
private val b = StringBuilder()
private var personA = 0
private var personB = 0
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    linked = Array(n + 1) { ArrayList() }
    IntStream.range(1, n + 1).forEach { linked[it] = ArrayList() }
    isVisited = BooleanArray(n + 1)
    val t = StringTokenizer(r.readLine())
    personA = t.nextToken().toInt()
    personB = t.nextToken().toInt()
    repeat(r.readLine().toInt()) {
        val (x, y) = r.readLine().split(" ").map { it.toInt() }
        linked[x].add(y)
        linked[y].add(x)
    }
    dfs(personA)
    if (b.isEmpty()) w.append("-1").close()
    else w.append(b).close()
}
private fun dfs(people: Int) {
    if (isVisited[people]) return
    isVisited[people] = true
    if (people == personB) b.append(count)
    count++
    linked[people].forEach { nextPeople -> dfs(nextPeople) }
    count--
}